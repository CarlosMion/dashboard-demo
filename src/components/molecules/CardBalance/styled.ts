"use client";

import { styled } from "@mui/material/styles";
import Typography from "@mui/material/Typography";
import Stack from "@mui/material/Stack";
import useMediaQuery from "@mui/material/useMediaQuery";
import { CreditCardVariant } from "../Card/types";

interface TitleProps {
  contrast: CreditCardVariant;
}

export const Container = styled(Stack)({
  display: "flex",
  flexDirection: "column",
  justifyContent: "center",
  alignItems: "start",
});

export const Title = styled(Typography)<TitleProps>(({ theme, contrast }) => ({
  color:
    contrast === "dark"
      ? theme.palette.common.white
      : theme.palette.text.secondary,
}));

export const Value = styled(Typography)<TitleProps>(({ theme, contrast }) => ({
  color:
    contrast === "dark" ? theme.palette.common.white : theme.palette.text.title,
  fontSize: useMediaQuery(theme.breakpoints.down("md")) ? "1rem" : "1.5rem",
  fontWeight: 500,
}));
