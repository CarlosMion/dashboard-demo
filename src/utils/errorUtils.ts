export type ServerError = {
  error: string;
  statusCode?: number;
};

export const isServerErrorWithStatusCode = (
  result: object
): result is ServerError => {
  return "error" in result && "statusCode" in result;
};

export const isServerError = (result: unknown): result is ServerError => {
  return typeof result === "object" && result !== null && "error" in result;
};

type ErrorWithMessage = {
  message: string;
};

export const isErrorWithMessage = (
  error: unknown
): error is ErrorWithMessage => {
  return (
    typeof error === "object" &&
    error !== null &&
    "message" in error &&
    typeof (error as Record<string, unknown>).message === "string"
  );
};

export const toErrorWithMessage = (maybeError: unknown): ErrorWithMessage => {
  if (isErrorWithMessage(maybeError)) return maybeError;

  try {
    return new Error(JSON.stringify(maybeError));
  } catch {
    // fallback in case there's an error stringifying the maybeError
    // like with circular references for example.
    return new Error(String(maybeError));
  }
};

export const getErrorMessage = (error: unknown): string => {
  return toErrorWithMessage(error).message;
};

export class ApiError extends Error {
  statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = "ApiError";
    this.statusCode = statusCode || 500;
  }
}
