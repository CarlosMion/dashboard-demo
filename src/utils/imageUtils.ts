export const dataURItoBlob = (dataURI: string, fileType: string): Blob => {
  const byteString = atob(dataURI.split(',')[1]);
  const arrayBuffer = new ArrayBuffer(byteString.length);
  const intArray = new Uint8Array(arrayBuffer);

  for (let i = 0; i < byteString.length; i++) {
    intArray[i] = byteString.charCodeAt(i);
  }

  return new Blob([arrayBuffer], { type: fileType });
};

export const dataURItoFile = (
  dataURI: string,
  fileName: string,
  fileType: string
): File => {
  const blob = dataURItoBlob(dataURI, fileType);
  return new File([blob], fileName, { type: blob.type });
};

export const blobToFile = (blob: Blob, fileName: string): File => {
  return new File([blob], fileName, { type: blob.type });
};

export const getImageDimensions = (
  src: string
): Promise<{ width: number; height: number } | null> => {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = src;

    img.onload = () =>
      resolve(
        img.width && img.height
          ? { width: img.width, height: img.height }
          : null
      );
    img.onerror = (error) => reject(error);
  });
};

export const isImageDimensionWithinThreshold = (
  actual: number,
  ideal: number,
  threshold: number
): boolean => {
  const isWithinThreshold = actual >= ideal * (1 - threshold);
  return isWithinThreshold;
};
