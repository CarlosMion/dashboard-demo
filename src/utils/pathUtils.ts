export function toCamelCase(str?: string): string | undefined {
  return str?.replace(/-([a-z])/g, (match, letter) => letter.toUpperCase());
}
