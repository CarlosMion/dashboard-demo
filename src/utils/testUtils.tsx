import { render } from "@testing-library/react";
import { ReactNode } from "react";
import AppProviders from "@/providers/AppProviders/AppProviders";

import messages from "@/i18n/messages/en.json";

const customRender = (children: ReactNode) =>
  render(
    <AppProviders
      locale={"en"}
      messages={messages}
      userAgent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    >
      {children}
    </AppProviders>
  );

export * from "@testing-library/react";

export { customRender as render };
