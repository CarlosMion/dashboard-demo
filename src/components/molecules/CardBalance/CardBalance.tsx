import { CreditCardVariant } from "../Card/types";
import { Value, Container, Title } from "./styled";

interface CardBalanceProps {
  title: string;
  value: string;
  variant?: CreditCardVariant;
}

export default function CardBalance({
  title,
  value,
  variant = "dark",
}: CardBalanceProps) {
  return (
    <Container>
      <Title variant="subtitle2" contrast={variant}>
        {title}
      </Title>
      <Value variant="subtitle2" contrast={variant}>
        {value}
      </Value>
    </Container>
  );
}
