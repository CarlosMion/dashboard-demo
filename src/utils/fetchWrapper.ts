import { isServerError } from "./errorUtils";

const fetcher = async <T>(
  info: RequestInfo | URL,
  init?: RequestInit
): Promise<T> => {
  let result;
  try {
    const injectedHeaders: Record<string, string> = {};
    if (typeof window === "undefined") {
      const headers = await import("next/headers");
      (await headers.headers()).forEach((value, key) => {
        injectedHeaders[key] = value;
      });
    }

    const buildHeaders: Record<string, string> = {};
    let url = info;
    if (typeof info === "string") {
      url = `${process.env.NEXT_PUBLIC_WEBAPP_API_URL}/${info}`;
    }
    const reqHeaders = {
      "Content-Type": "application/json",
      ...init?.headers,
      ...injectedHeaders,
      ...buildHeaders,
    };
    result = await fetch(url, {
      ...init,
      headers: reqHeaders,
    });
    if (!result.ok) {
      throw new Error("Failed to fetch");
    }
    let parsedResult;
    try {
      parsedResult = await result.json();
    } catch (error) {
      console.log("Failed to parse response, probably not JSON", result, error);
      return result as T;
    }
    if (isServerError(parsedResult)) {
      throw new Error(parsedResult.error);
    }
    return parsedResult as T;
  } catch (error) {
    throw error;
  }
};

export default fetcher;
