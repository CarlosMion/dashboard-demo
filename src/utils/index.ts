import { isEdge } from "./userAgentUtils";
import { maskCardNumber } from "./stringUtils";
import { toCamelCase } from "./pathUtils";
import fetcher from "./fetchWrapper";
import {
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
} from "./errorUtils";
import {
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
} from "./imageUtils";
import {
  createImage,
  getRadianAngle,
  rotateSize,
  getCroppedImg,
} from "./cropImageUtils";
export {
  fetcher,
  isEdge,
  toCamelCase,
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
  getCroppedImg,
  rotateSize,
  getRadianAngle,
  createImage,
  maskCardNumber,
};
