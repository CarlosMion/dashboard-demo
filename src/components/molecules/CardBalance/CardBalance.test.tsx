import { render, screen } from "@/utils/testUtils";
import CardBalance from "./CardBalance";

describe("CardBalance", () => {
  const defaultProps = {
    title: "Available Balance",
    value: "$1,000",
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<CardBalance {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("displays the title", () => {
    render(<CardBalance {...defaultProps} />);
    expect(screen.getByText("Available Balance")).toBeInTheDocument();
  });

  it("displays the value", () => {
    render(<CardBalance {...defaultProps} />);
    expect(screen.getByText("$1,000")).toBeInTheDocument();
  });

  it("applies dark variant by default", () => {
    const { container } = render(<CardBalance {...defaultProps} />);
    expect(container.firstChild).toHaveStyle({ color: expect.any(String) });
  });

  it("applies light variant when specified", () => {
    const { container } = render(
      <CardBalance {...defaultProps} variant="light" />
    );
    expect(container.firstChild).toHaveStyle({ color: expect.any(String) });
  });
});
