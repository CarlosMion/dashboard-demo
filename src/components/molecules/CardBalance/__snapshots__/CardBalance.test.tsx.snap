// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardBalance renders correctly and matches snapshot 1`] = `
<div
  class="MuiStack-root css-1e6kwg5-MuiStack-root"
>
  <h6
    class="MuiTypography-root MuiTypography-subtitle2 css-1mlb6s-MuiTypography-root"
    contrast="dark"
  >
    Available Balance
  </h6>
  <h6
    class="MuiTypography-root MuiTypography-subtitle2 css-lgs5qv-MuiTypography-root"
    contrast="dark"
  >
    $1,000
  </h6>
</div>
`;
